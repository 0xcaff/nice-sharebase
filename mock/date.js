// Sets any calls to date to return a constant time with a lot of zeros.
const constantDate = new Date(1500000000000);

global.Date.now = () => constantDate;