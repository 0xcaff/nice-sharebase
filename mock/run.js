require('babel-register');
require("regenerator-runtime/runtime");

const app = require('./server');
app.listen(30101, '0.0.0.0', function() { console.log('started mock server on port 30101') });
