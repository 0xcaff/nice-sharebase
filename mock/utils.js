import fs from 'fs';
import url, { URLSearchParams } from 'url';
import path from 'path';

// Returns the result of the first value in elems send through predicate which didn't throw an error.
export async function first(elems, predicate) {
	for (let elem of elems) {
		const { result, error } = await tryIt(_ => predicate(elem));
		if (!error) {
			return result;
		}
	}
}

// Wraps fs.readFile to use promises.
export const readFile = (...args) =>
  new Promise((resolve, reject) =>
    fs.readFile(...args, (err, data) => {
      if (err) { reject(err) } else { resolve(data) }
    })
  );

// Async returns an object with an error if one was caught. Otherwise returns the
// result.
export const tryIt = async (it) => {
  try {
    return {result: await it()};
  } catch(e) {
    return {error: e};
  }
};

// If a folder is requested with embeddings added, add we add them here.
export async function embedIfNeeded(folder, location) {
  const params = new URLSearchParams(location.search);
  const embed = params.get('embed');

  if (!embed) {
    return folder;
  }

  const embeds = embed.split(',');
  folder = JSON.parse(folder);

  const id = folder['FolderId'];
  if (embeds.includes('f')) {
    // try to embed folders
    const file = path.resolve(__dirname, `api/folders/${id}/folders.json`);

    const { result, error } = await tryIt(_ => readFile(file));
    if (!error && result) {
      const folders = JSON.parse(result);
      folder["Embedded"]["Folders"].push(...folders);
    }
  }

  // try to get folders
  if (embeds.includes('d')) {
    // try to embed documents
    const file = path.resolve(__dirname, `api/folders/${id}/documents.json`);

    const { result, error } = await tryIt(_ => readFile(file));
    if (!error && result) {
      const documents = JSON.parse(result);
      folder["Embedded"]["Documents"].push(...documents);
    }
  }

  return JSON.stringify(folder);
};

export async function handle(p) {
  const location = url.parse(p);
  p = location.pathname.slice(1);

	const paths = [`${p}.json`, `${p}/index.json`]
 		.map(raw => path.resolve(__dirname, raw));

  const result = await first(paths, path => readFile(path));
  if (result !== undefined) {
    const status = 200;
    const message = await embedIfNeeded(result, location);
    const headers = {'Content-Type': 'application/json'};

    return { status, message, headers };
  } else {
    const status = 404;
		const pre = `Path not found for ${p}. Tried:`;
		const tried = paths.map(path => '\t' + path).join('\n');
    const message = pre + '\n' + tried;

    return { status, message };
  }
}
