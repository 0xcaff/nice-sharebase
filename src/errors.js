import { GraphQLError } from 'graphql';

// An error which occurs when making a request fails.
export class FetchError extends GraphQLError {
  constructor(resp, body) {
    super(`Request Failed. Status: ${resp.status} - ${resp.statusText}. Body: ${body}`);

    this.body = body;
    this.resp = resp;
  }
}

export const throwOnFail = async (resp) => {
  if (resp.ok) {
    return resp;
  } else {
    throw new FetchError(resp, await resp.text());
  }
}