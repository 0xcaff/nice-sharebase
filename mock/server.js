// Creates an express server to serve the mock data.
import path from 'path';
import express from 'express';
import { readFile, first } from './utils';

const app = express();

// handle all paths dynamically
app.get('*', (req, res, next) => {
  handle(req, res)
    .then(_ => next());
});

export default app;

const handle = async (req, res) => {
  const rPath = req.path.slice(1);

  // try resolving the url. Examples:
  //
  // /api/libraries
  // /api/libraries/{libraryid}
  // /api/libraries/{libraryid}/folders
  //
  // /api/folders/{folderid}/documents
  // /api/folders/{folderid}/folders

  const paths = [`${rPath}.json`, `${rPath}/index.json`]
    .map(raw => path.resolve(__dirname, raw));

  const result = await first(paths, path => readFile(path));
  if (result !== undefined) {
    res
      .set('Content-Type', 'application/json')
      .send(result);
  } else {
    const message = `Path not found for ${rPath}. Tried:`;
    const tried = paths.map(path => '\t' + path).join('\n');

    res
      .status(404)
      .send(message + '\n' + tried);
  }
};

