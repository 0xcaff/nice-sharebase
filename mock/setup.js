// Sets up nock to respond to requests to the mock endpoint.
import nock from 'nock';

import { handle } from './utils';

export const ENDPOINT = 'http://mock.com/';

export const logs = [];

global.beforeEach && beforeEach(() => logs.length = 0);

const setup = app => app
  .get(() => true)
  .reply(function(url, body, cb) {
    h(url)
      .then((...args) => {
        // setup for next request
        setup(app);
        cb(null, ...args);
      })
      .catch(err => cb(err));
  });

const app = nock(ENDPOINT);
setup(app);

async function h(path) {
  log(path);
  const { status, message, headers } = await handle(path);
  return [status, message, headers];
}

const log = (path) => logs.push(path);

