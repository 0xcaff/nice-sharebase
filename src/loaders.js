import { BatchLoader, Loader } from './loader';
import { filter } from './utils';

// Called usually once per request to create the caches responsible for holding
// on to data.
export const create = ({ network }) => ({
  library: Library({ network }),
  libraryFolders: LibraryFolders({ network }),
  folder: Folder({ network }),
  document: Document({ network }),
});

const Library = ({ network }) => new LibraryLoader({ network });

class LibraryLoader extends BatchLoader {
  constructor({ network, autoDispatch }) {
    super({ autoDispatch });
    this.network = network;
  }

  async batchLoad(queue) {
    const { network } = this;

    // find minimum required work, we will only have one qualifying promise
    // because of the cache
    const gettingAll = queue.length > 1 ||
      queue.some(({ extras }) => extras === 'all');

    if (gettingAll) {
      // if we fail here, the entire batch will fail transiently
      const all = await network.request('libraries');

      queue.forEach(({ id, extras, resolve, reject }) => {
        if (extras === 'all') {
          resolve(all);
        }

        const lib = all.find(lib => lib['LibraryId'] == id);
        if (!lib) {
          reject(new TypeError(`The library with id ${id} can't be found.`));
        }

        resolve(lib);
      });
    } else {
      // resolve queue one by one
      return await Promise.all(
        queue.map(async ({ id, resolve, reject }) => {
          try {
            const body = await network.request(`libraries/${id}`);
            resolve(body);
          } catch(e) {
            reject(e);
          }
        })
      );
    }
  }
}

const LibraryFolders = ({ network }) => new Loader({
  fetchFn: async (id) => await network.request(`libraries/${id}/folders`),
});

const Folder = ({ network }) => new FolderLoader({ network });

export class FolderLoader extends BatchLoader {
  constructor({ network, autoDispatch }) {
    super({ autoDispatch });
    this.network = network;
  }

  getFromCache(id, { folders: needFolders, documents: needDocs }) {
    const cached = this.cache.get(id);
    if (!cached) {
      return null;
    }

    const { extras, promise } = cached;
    const { folders: hasFolders, documents: hasDocs } = extras;

    // ensure that the extras for the cached request are better than or equal to
    // the one requested
    if (needFolders && !hasFolders) {
      return null;
    }

    if (needDocs && !hasDocs) {
      return null;
    }

    return promise;
  }

  async batchLoad(queue) {
    const { network } = this;

    // reduce into minimum required requests
    const reqs = queue.reduce(
      (acc, { id, extras: rExtras, resolve, reject }) => {
        const { promiseFuncs = [], extras = {} } = (acc[id] || {});

        // update
        promiseFuncs.push({ resolve, reject });
        Object.assign(extras, filter(rExtras));

        acc[id] = { promiseFuncs, extras };
        return acc;
      }, {});

    return await Promise.all(
      Object.entries(reqs)
        .map(async ([id, { promiseFuncs, extras: { folders, documents }} ]) => {
          try {
            const paramsString =
              Object.keys(filter({ 'f': folders, 'd': documents }))
                .join();

            const path = `folders/${id}` + (paramsString && `?embed=${paramsString}`);
            const body = await network.request(path);

            promiseFuncs.forEach(({ resolve }) => resolve(body));
          } catch(e) {
            promiseFuncs.forEach(({ reject }) => reject(e));
          }
        })
    );
  }
}

const Document = ({ network }) => new Loader({
  fetchFn: async (id) => await network.request(`documents/${id}`),
});