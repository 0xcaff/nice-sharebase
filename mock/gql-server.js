import gql from 'express-graphql';
import express from 'express';
import url from 'url';
import { schema, createContext } from '../index.js';

import { ENDPOINT, logs } from './setup';

const app = express();

app.use('/graphql', gql(
  async (req, res, params) => {
    const base = url.resolve(ENDPOINT, 'api/');
    const context = createContext({ base });

    return {
      schema,
      context,
      extensions: (document, variables, operationName, result) => {
        const l = logs.slice();
        logs.length = 0;
        return l;
      },
      graphiql: true,
      pretty: true,
    };
  }
));

app.listen(3010, '0.0.0.0', _ => console.log('listening on 0.0.0.0:3010'));
