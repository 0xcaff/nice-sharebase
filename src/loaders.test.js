import url from 'url';
import 'isomorphic-fetch';

import { FolderLoader } from './loaders';
import { throwOnFail } from './errors';
import { setup, DEFAULT_ENDPOINT } from '../mock/setup';

setup({ authEnabled: false });
const base = url.resolve(DEFAULT_ENDPOINT, 'api/');

const logs = [];
beforeEach(() => logs.length = 0);

// A network mock without authentication.
const network = {
  request: async path => {
    const fullPath = url.resolve(base, path);
    logs.push(fullPath);
    const resp = await throwOnFail(await fetch(fullPath));
    const body = await resp.json();

    return body;
  },
};

it('should batch loads to folders and documents in a folder', async () => {
  const loader = new FolderLoader({ network, autoDispatch: false });

  const responses = [
    loader.load(6890),
    loader.load(6890, { folders: true }),
    loader.load(6890, { documents: true }),
  ];

  await loader.dispatch();
  const data = await Promise.all(responses);

  expect(data[0]).toEqual(data[1]);
  expect(data[1]).toEqual(data[2]);
  expect(data[0]['Embedded']).toMatchSnapshot();
  expect(logs).toMatchSnapshot();
});

it('should cache loads for a folder between dispatches', async () => {
  const loader = new FolderLoader({ network, autoDispatch: false });

  const first = loader.load(6890);
  await loader.dispatch();

  const second = loader.load(6890);
  await loader.dispatch();

  expect(await first).toMatchSnapshot();
  expect(await first).toEqual(await second);
  expect(logs).toMatchSnapshot();
});

it('should automatically execute loader.dispatch', async () => {
  const loader  = new FolderLoader({ network });

  const first = await loader.load(6890);
  expect(first).toMatchSnapshot();
  expect(logs).toMatchSnapshot();
});

// TODO: test multiple case
// TODO: test single failure case
// TODO: test mass failure case