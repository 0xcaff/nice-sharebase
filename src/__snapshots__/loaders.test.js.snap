// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should automatically execute loader.dispatch 1`] = `
Object {
  "Embedded": Object {
    "Documents": Array [],
    "Folders": Array [],
  },
  "FolderId": 6890,
  "FolderName": "Suit",
  "Links": Object {
    "Documents": "https://app.sharebase.com/ShareBaseAPI/api/folders/6890/documents",
    "Folders": "https://app.sharebase.com/ShareBaseAPI/api/folders/6890/folders",
    "Self": "https://app.sharebase.com/ShareBaseAPI/api/folders/6890",
    "Shares": "https://app.sharebase.com/ShareBaseAPI/api/folders/6890/share",
  },
}
`;

exports[`should automatically execute loader.dispatch 2`] = `
Array [
  "http://mock.com/api/folders/6890",
]
`;

exports[`should batch loads to folders and documents in a folder 1`] = `
Object {
  "Documents": Array [
    Object {
      "DateModified": "2016-02-08T19:25:51.667Z",
      "DocumentId": 19749,
      "DocumentName": "Armor.cad",
      "Links": Object {
        "Content": "https://app.sharebase.com/ShareBaseAPI/api/documents/19749/content",
        "Self": "https://app.sharebase.com/ShareBaseAPI/api/documents/19749",
      },
    },
    Object {
      "DateModified": "2016-02-08T19:26:01.167Z",
      "DocumentId": 19779,
      "DocumentName": "Pitch.docx",
      "Links": Object {
        "Content": "https://app.sharebase.com/ShareBaseAPI/api/documents/19779/content",
        "Self": "https://app.sharebase.com/ShareBaseAPI/api/documents/19779",
      },
    },
  ],
  "Folders": Array [
    Object {
      "Embedded": Object {
        "Documents": Array [],
        "Folders": Array [],
      },
      "FolderId": 2034,
      "FolderName": "Hot Rod Red",
      "Links": Object {
        "Documents": "https://app.sharebase.com/ShareBaseAPI/api/folders/2034/documents",
        "Folders": "https://app.sharebase.com/ShareBaseAPI/api/folders/2034/folders",
        "Self": "https://app.sharebase.com/ShareBaseAPI/api/folders/2034",
        "Shares": "https://app.sharebase.com/ShareBaseAPI/api/folders/2034/share",
      },
    },
  ],
}
`;

exports[`should batch loads to folders and documents in a folder 2`] = `
Array [
  "http://mock.com/api/folders/6890?embed=f,d",
]
`;

exports[`should cache loads for a folder between dispatches 1`] = `
Object {
  "Embedded": Object {
    "Documents": Array [],
    "Folders": Array [],
  },
  "FolderId": 6890,
  "FolderName": "Suit",
  "Links": Object {
    "Documents": "https://app.sharebase.com/ShareBaseAPI/api/folders/6890/documents",
    "Folders": "https://app.sharebase.com/ShareBaseAPI/api/folders/6890/folders",
    "Self": "https://app.sharebase.com/ShareBaseAPI/api/folders/6890",
    "Shares": "https://app.sharebase.com/ShareBaseAPI/api/folders/6890/share",
  },
}
`;

exports[`should cache loads for a folder between dispatches 2`] = `
Array [
  "http://mock.com/api/folders/6890",
]
`;
